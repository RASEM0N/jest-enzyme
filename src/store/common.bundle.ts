import { createAction, createSlice } from '@reduxjs/toolkit'
import { all, takeEvery, delay, put } from 'redux-saga/effects'

// -------------------
// ----- TOOLKIT -----
// -------------------
const commonSlice = createSlice({
    name: 'common',
    initialState: {
        number: 0,
    },
    reducers: {
        increment: (state) => {
            state.number++
        },
        decrement: (state) => {
            state.number--
        },
    },
})

export default commonSlice
export const { decrement, increment } = commonSlice.actions
export const commonReducer = commonSlice.reducer
export const asyncIncrement = createAction('async:increment')
export const asyncDecrement = createAction('async:decrement')

// -------------------
// ------ SAGA -------
// -------------------

export const decrementStore = function* () {
    yield delay(3000)
    yield put(decrement())
}

export const incrementStore = function* () {
    yield delay(3000)
    yield put(increment())
}

export const watchChangeAction = function* () {
    yield takeEvery(asyncDecrement.type, decrementStore)
    yield takeEvery(asyncIncrement.type, incrementStore)
}

export const commonRootSaga = function* () {
    yield all([watchChangeAction()])
}
