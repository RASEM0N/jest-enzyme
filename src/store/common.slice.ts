import { createSlice } from '@reduxjs/toolkit'

const commonSlice = createSlice({
    name: 'common',
    initialState: {
        number: 0,
    },
    reducers: {
        increment: (state) => {
            state.number++
        },
        decrement: (state) => {
            state.number--
        },
    },
})

export const { decrement, increment } = commonSlice.actions
export default commonSlice.reducer
