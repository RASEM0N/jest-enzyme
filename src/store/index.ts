import { configureStore } from '@reduxjs/toolkit'
import { useDispatch } from 'react-redux'
import createSagaMiddleware from 'redux-saga'
import { all } from 'redux-saga/effects'
import { commonReducer, commonRootSaga } from './common.bundle'

// -------------------
// ------ SAGA -------
// -------------------
const sagaMiddleware = createSagaMiddleware()
const rootSaga = function* () {
    yield all([commonRootSaga()])
}

// -------------------
// ----- STORE -------
// -------------------

export const store = configureStore({
    devTools: true,
    reducer: {
        common: commonReducer,
    },
    middleware: [sagaMiddleware],
})

sagaMiddleware.run(rootSaga)

// -------------------
// ----- COMMON ------
// -------------------
export const useAppDispatch = () => {
    return useDispatch<AppDispatch>()
}
export type AppStoreType = typeof store
export type AppDispatch = typeof store.dispatch
export type StateType = ReturnType<typeof store.getState>
