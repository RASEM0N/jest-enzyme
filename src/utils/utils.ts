import axios from 'axios'

// eslint-disable-next-line @typescript-eslint/ban-types
export const forEach = (items: Array<number>, callback: Function) => {
    for (let index = 0; index < items.length; index++) {
        callback(items[index])
    }
}

export class Users {
    static all(): Promise<any> {
        return axios.get('/users.json').then((resp) => {
            return resp.data
        })
    }
}
