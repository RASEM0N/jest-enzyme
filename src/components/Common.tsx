import React from 'react'
import { StateType, useAppDispatch } from '../store'
import { useSelector } from 'react-redux'
import {
    decrement,
    increment,
    asyncIncrement,
    asyncDecrement,
} from '../store/common.bundle'

const Common: React.FunctionComponent = () => {
    const dispatch = useAppDispatch()
    const numberFromState = useSelector<StateType>((state) => {
        return state.common.number
    })

    const incr = () => {
        dispatch(increment())
    }

    const decr = () => {
        dispatch(decrement())
    }

    const asyncIncr = () => {
        dispatch(asyncIncrement())
    }

    const asyncDecr = () => {
        dispatch(asyncDecrement())
    }

    return (
        <React.Fragment>
            <h2>{numberFromState}</h2>
            <button onClick={incr}>+1</button>
            <button onClick={decr}>-1</button>
            <button onClick={asyncIncr}>async +1</button>
            <button onClick={asyncDecr}>async -1</button>
        </React.Fragment>
    )
}

export default Common
