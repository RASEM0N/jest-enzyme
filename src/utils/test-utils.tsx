import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import { EnhancedStore } from '@reduxjs/toolkit'
import { AppDispatch, StateType } from '../store'

export const renderWithRedux = (Component: JSX.Element, store: EnhancedStore) => {
    const component = mount(<Provider store={store}>{Component}</Provider>)
    const state = store.getState() as StateType
    const dispatch = store.dispatch as AppDispatch

    return {
        component,
        state,
        dispatch,
    }
}
