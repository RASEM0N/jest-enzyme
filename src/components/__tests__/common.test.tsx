import React from 'react'
import Common from '../Common'
import { mount } from 'enzyme'
import { renderWithRedux } from '../../utils/test-utils'
import { store } from '../../store'

describe('common.tsx', () => {
    describe('with redux', () => {
        let myStore

        beforeEach(() => {
            myStore = store
        })

        it('snapshots', () => {
            const { component } = renderWithRedux(<Common />, myStore)

            expect(component.debug()).toMatchSnapshot()
        })

        it('click on a incr and decr button', () => {
            const { component } = renderWithRedux(<Common />, myStore)

            const btnIncr = component.find('#incr')
            btnIncr.simulate('click')
            expect(myStore.getState().common.number).toBe(1)

            const btnDecr = component.find('#decr')
            btnDecr.simulate('click')
            expect(myStore.getState().common.number).toBe(-1)
        })
    })
})
