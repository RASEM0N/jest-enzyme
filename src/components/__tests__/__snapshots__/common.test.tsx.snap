// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common.tsx with redux snapshots 1`] = `
"<Provider store={{...}}>
  <Common>
    <h2>
      0
    </h2>
    <button id=\\"incr\\" onClick={[Function: incr]}>
      +1
    </button>
    <button id=\\"decr\\" onClick={[Function: decr]}>
      -1
    </button>
    <button id=\\"asyncIncr\\" onClick={[Function: asyncIncr]}>
      async +1
    </button>
    <button id=\\"asyncDecr\\" onClick={[Function: asyncDecr]}>
      async -1
    </button>
  </Common>
</Provider>"
`;
